code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,489 @@
+    The argument passed is a list of blacklisted traits you don't your artifact to have, allowing
+    for a defenition of artifact types.
+    The process also generates some partial hints, like a touch description and science-glasses desrection(special_desc)
+    malf is an option to nab a malfunction trait on init. See URANIUM types.
+*/
+/obj/item/xenoartifact/proc/generate_traits(var/list/blacklist_traits, malf = FALSE)
+    var/datum/xenoartifact_trait/new_trait
+
+    while(!(new_trait) || (new_trait in blacklist_traits)) //Activator
+        new_trait = pick(subtypesof(/datum/xenoartifact_trait/activator))
+    traits[1] = new new_trait
+    blacklist_traits += list(traits[1].blacklist_traits)
+    blacklist_traits += list(new_trait)
+    new_trait = null
+    special_desc = traits[1].desc ? "[special_desc] [traits[1].desc]" : "[special_desc]"
+
+    var/minor_desc
+    for(var/X in 2 to 4)//Minors
+        while((!new_trait) || (new_trait in blacklist_traits))
+            new_trait = pick(subtypesof(/datum/xenoartifact_trait/minor))
+        traits[X] = new new_trait
+        blacklist_traits += list(traits[X].blacklist_traits)
+        blacklist_traits += list(new_trait)
+        new_trait = null
+        if(traits[X].on_touch(src, src) && !touch_desc)
+            touch_desc = traits[X]
+        if(!(minor_desc) && traits[X].desc)
+            minor_desc = traits[X].desc
+    special_desc = minor_desc ? "[special_desc] [minor_desc] material." : "[special_desc] material."
+
+
+    while(!(new_trait) || (new_trait in blacklist_traits))//Major
+        new_trait = pick(subtypesof(/datum/xenoartifact_trait/major))
+    traits[5] = new new_trait
+    blacklist_traits += list(traits[5].blacklist_traits)
+    blacklist_traits += list(new_trait)
+    new_trait = null
+    special_desc = traits[5].desc ? "[special_desc] The shape is [traits[5].desc]." : "[special_desc]"
+
+    charge_req = 10*rand(1, 10) //This is here just in-case I decide to change how this works.
+
+    if(!malf)
+        return
+    while(!new_trait||get_trait(new_trait, blacklist_traits))//Malfunctions
+        new_trait = pick(subtypesof(/datum/xenoartifact_trait/malfunction))
+    traits[6] = new new_trait
+    blacklist_traits += list(traits[6].blacklist_traits)
+    new_trait = null
+    
+/obj/item/xenoartifact/proc/get_proximity(range) //Gets a singular bam beano
+    for(var/mob/living/M in range(range, get_turf(src)))
+        return process_target(M)
+    if(isliving(loc))
+        return loc
+
+/obj/item/xenoartifact/proc/get_trait(typepath) //Returns the desired trait and it's values if it's in the artifact's
+    for(var/datum/xenoartifact_trait/T in traits)
+        if(istype(T, typepath))
+            return T
+    return FALSE
+
+/obj/item/xenoartifact/proc/generate_icon(var/icn, var/icnst = "", colour) //Add extra icon overlays
+    icon_overlay = mutable_appearance(icn, icnst)
+    icon_overlay.layer = FLOAT_LAYER
```

Comment:
Will this get screwy if the artifact is placed on a turf with an item between the 2 layers? Perhaps it should use the artifacts layer + 0.1 or some really small value to prevent that happening

Reply:
(Overlay might prevent this though, I'm not sure)